kratos/processes/apply_periodic_boundary_condition_process.cpp

```diff
@@ -149,92 +150,125 @@ template <int TDim>
 void ApplyPeriodicConditionProcess::ApplyConstraintsForPeriodicConditions()
 {
     const auto timer = BuiltinTimer();
-    const int num_vars = mParameters["variable_names"].size();
+    const unsigned int num_vars = mParameters["variable_names"].size();
     BinBasedFastPointLocatorConditions<TDim> bin_based_point_locator(mrMasterModelPart);
     bin_based_point_locator.UpdateSearchDatabase();
 
-    IndexType num_slaves_found = 0;
+    // Define auxiliary functions
+    const std::function<void(NodeType&, const GeometryType&, const VectorType&, const std::vector<const VariableType*>&)> function_for_vector_variable =
+    [this](NodeType& rSlaveNode, const GeometryType& rHostedGeometry, const VectorType& rWeights, const std::vector<const VariableType*>& rVars) {ConstraintSlaveNodeWithConditionForVectorVariable<TDim>(rSlaveNode, rHostedGeometry, rWeights, rVars);};
+    const std::function<void(NodeType&, const GeometryType&, const VectorType&, const std::vector<const VariableType*>&)> function_for_scalar_variable =
+    [this](NodeType& rSlaveNode, const GeometryType& rHostedGeometry, const VectorType& rWeights, const std::vector<const VariableType*>& rVars) {ConstraintSlaveNodeWithConditionForScalarVariable<TDim>(rSlaveNode, rHostedGeometry, rWeights, rVars);};
+
+    // Fill an auxiliary vector with the functions
+    std::vector<const std::function<void(NodeType&, const GeometryType&, const VectorType&, const std::vector<const VariableType*>&)>*> functions_required(num_vars, nullptr);
```

Comment:
If you have a setup where you need to allocate some objects that should be in an invalid / uninitialized state before they get assigned, you can use [```std::optional```](https://en.cppreference.com/w/cpp/utility/optional) instead of raw pointers and ```nullptr```. 

```std::optional``` is stored on the stack and only introduces one additional byte (a boolean variable that indicates whether the stored object has a valid value), so it's a minimal overhead but much safer than gambling with ```nullptr```s.

Reply:
Sorry, I don't get how to use optional in this situation https://www.cppstories.com/2018/05/using-optional/

Reply:
Example:
```diff
-    std::vector<const MyFunction*> functions_required(num_vars, nullptr);
+    std::vector<std::optional<std::reference_wrapper<const MyFunction>>> functions_required(num_vars);
    std::vector<std::vector<const VariableType*>> variables_vector(num_vars, std::vector<const VariableType*>());
    for (unsigned int j = 0; j < num_vars; j++) {
        const std::string& r_var_name = mParameters["variable_names"][j].GetString();
        if (KratosComponents<Variable<array_1d<double, 3>>>::Has(r_var_name)) {
-            functions_required[j] = &function_for_vector_variable;
+            functions_required[j] = function_for_vector_variable;
            variables_vector[j].push_back(&KratosComponents<VariableType>::Get(r_var_name + "_X"));
            variables_vector[j].push_back(&KratosComponents<VariableType>::Get(r_var_name + "_Y"));
            variables_vector[j].push_back(&KratosComponents<VariableType>::Get(r_var_name + "_Z"));
        } else if (KratosComponents<VariableType>::Has(r_var_name)) {
-            functions_required[j] = &function_for_scalar_variable;
+            functions_required[j] = function_for_scalar_variable;
```
but now that I'm looking at it, you could also just use the classic ```reserve``` and ```push_back``` combo in this case, because you're just iterating contiguously.